cdap-watchdog/src/main/java/co/cask/cdap/logging/appender/loader/LogAppenderLoaderService.java

```diff
@@ -0,0 +1,137 @@
+/*
+ * Copyright © 2019 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.logging.appender.loader;
+
+import ch.qos.logback.classic.Logger;
+import ch.qos.logback.classic.LoggerContext;
+import ch.qos.logback.classic.spi.ILoggingEvent;
+import ch.qos.logback.core.Appender;
+import ch.qos.logback.core.status.OnConsoleStatusListener;
+import ch.qos.logback.core.status.StatusManager;
+import co.cask.cdap.common.conf.CConfiguration;
+import co.cask.cdap.common.conf.Constants;
+import com.google.common.collect.Iterators;
+import com.google.common.util.concurrent.AbstractIdleService;
+import com.google.inject.Inject;
+import org.elasticsearch.common.Strings;
+import org.slf4j.ILoggerFactory;
+import org.slf4j.LoggerFactory;
+
+import java.util.Collections;
+import java.util.Map;
+
+/**
+ * Service that programmatically loads additional log appenders. It provides classloader isolation for additional
+ * appender.
+ */
+public class LogAppenderLoaderService extends AbstractIdleService {
+  private static final org.slf4j.Logger LOG = LoggerFactory.getLogger(LogAppenderLoaderService.class);
+  private final CConfiguration cConf;
+  private Appender<ILoggingEvent> logAppender;
+
+  @Inject
+  public LogAppenderLoaderService(CConfiguration cConf) {
+    this.cConf = cConf;
+  }
+
+  @Override
+  protected void startUp() throws Exception {
+    String provider = cConf.get(Constants.Logging.LOG_APPENDER_PROVIDER);
+    // If there is no provider, there is nothing to load. This can happen if no additional appenders are provided.
+    if (Strings.isNullOrEmpty(provider)) {
+      return;
+    }
+
+    logAppender = loadLogAppender(provider, cConf);
+    initialize(logAppender, cConf);
+
+    LOG.info("Log Appender {} is initialized and started.", provider);
+  }
+
+  @Override
+  protected void shutDown() throws Exception {
+    if (logAppender != null) {
+      logAppender.stop();
+    }
+  }
+
+  /**
+   * Loads log appender using extension loader.
+   */
+  @SuppressWarnings("unchecked")
```

Comment:
What is this for?

Reply:
This is because `new LogAppenderExtensionLoader(cConf).get(appenderProvider)` returns `Appender` instead of `Appender<ILoggingEvent>`

Reply:
You can make it return `Appender<ILoggingEvent>` by specifying the right generic type in `LogAppenderExtensionLoader`

Reply:
fixed.